upup/pkg/fi/cloudup/gce/tpm/gcetpmsigner/tpmauthenticator.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcetpmsigner
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/google/go-tpm-tools/client"
+	"k8s.io/klog/v2"
+	"k8s.io/kops/pkg/bootstrap"
+	gcetpm "k8s.io/kops/upup/pkg/fi/cloudup/gce/tpm"
+)
+
+type tpmAuthenticator struct {
+	projectID string
+	zone      string
+	instance  string
+}
+
+var _ bootstrap.Authenticator = &tpmAuthenticator{}
+
+func NewTPMAuthenticator() (bootstrap.Authenticator, error) {
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil, fmt.Errorf("error getting projectID from metadata: %w", err)
+	}
+	zone, err := metadata.Zone()
+	if err != nil {
+		return nil, fmt.Errorf("error getting zone from metadata: %w", err)
+	}
+	instance, err := metadata.InstanceName()
+	if err != nil {
+		return nil, fmt.Errorf("error getting instance from metadata: %w", err)
+	}
+
+	return &tpmAuthenticator{
+		projectID: projectID,
+		zone:      zone,
+		instance:  instance,
+	}, nil
+}
+
+func (a *tpmAuthenticator) CreateToken(body []byte) (string, error) {
+	requestHash := sha256.Sum256(body)
+
+	tpmStart := time.Now()
+
+	tpmDevice, err := openTPM()
+	if err != nil {
+		return "", fmt.Errorf("failed to open TPM: %w", err)
+	}
+	defer tpmDevice.Close()
+
+	key, err := client.GceAttestationKeyRSA(tpmDevice)
+	if err != nil {
+		return "", fmt.Errorf("failed to get GCE RSA attestation key from TPM: %w", err)
+	}
+	defer key.Close()
+
+	klog.V(2).Infof("attestation key is %v", debugToPEM(key.PublicKey()))
+
+	klog.Infof("TPM initialization took %v", time.Since(tpmStart))
+
+	token := &gcetpm.AuthToken{}
+
+	token.Data.GCPProjectID = a.projectID
+	token.Data.Zone = a.zone
+	token.Data.Instance = a.instance
+	token.Data.Timestamp = time.Now().Unix()
+	token.Data.Audience = gcetpm.AudienceNodeAuthentication
+	token.Data.RequestHash = requestHash[:]
+
+	payload, err := json.Marshal(token.Data)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal token data: %w", err)
+	}
+	signature, err := tpmSign(key, payload)
+	if err != nil {
+		return "", fmt.Errorf("failed to sign token data: %w", err)
+	}
+	token.Signature = signature
```

Comment:
This is a bit fragile in that it relies on the client and server both serializing JSON the same way. When using a serialization format, such as JSON, which allows multiple possible encodings of the same thing it is better to transmit the precise byte sequence that got signed. Then again, for this particular case it is extremely likely for the client and server to be both using the same version of the same encoding library.

Reply:
Good point!  I guess it's not too hard to change it _now_ to be `payload []byte` and `signature []byte` (where it would be a pain later) and just do an additional JSON decode ... I'll make the change.